navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/instruction/DefaultInstructionEngine.java

```diff
@@ -0,0 +1,223 @@
+package com.mapbox.services.android.navigation.v5.instruction;
+
+import android.text.TextUtils;
+import android.util.SparseArray;
+
+import com.mapbox.services.android.navigation.v5.NavigationConstants;
+import com.mapbox.services.android.navigation.v5.RouteProgress;
+import com.mapbox.services.api.directions.v5.models.StepManeuver;
+import com.mapbox.services.api.utils.turf.TurfConstants;
+import com.mapbox.services.api.utils.turf.TurfHelpers;
+
+import java.text.DecimalFormat;
+import java.util.Locale;
+
+class DefaultInstructionEngine extends SparseArray<DefaultInstructionEngine.InstructionBuilder> {
```

Comment:
Since we have a set number of key-value pairs, should we define the sparse array size or is this not possible? See the [SparseArray constructor](https://developer.android.com/reference/android/util/SparseArray.html#SparseArray(int)).